File: Algorithms/Sorting/MergeSort/MergeSort.js
class MergeSort {

    static sort(arr) {
        const aux = arr.slice()
        MergeSort.sortHelper(arr, aux, 0, arr.length - 1)
    }

    static sortHelper(arr, aux, lo, hi) {
        if (hi <= lo) return
        const mid = Math.floor(lo + (hi - lo) / 2)
        MergeSort.sortHelper(arr, aux, lo, mid)
        MergeSort.sortHelper(arr, aux, mid + 1, hi)
        MergeSort.merge(arr, aux, lo, mid, hi)
    }

    static merge(arr, aux, lo, mid, hi) {
        for (let n = lo; n <= hi; n++) {
            aux[n] = arr[n]
        }
        let i = lo
        let j = mid + 1
        for (let k = lo; k <= hi; k++) {
            if (i > mid) {
                arr[k] = aux[j]
                j++
            } else if (j > hi) {
                arr[k] = aux[i]
                i++
            } else if (aux[i] < aux[j]) {
                arr[k] = aux[i]
                i++
            } else {
                arr[k] = aux[j]
                j++
            }
        }
    }

    static main() {
        const test1 = [6, -1, 66, -166, 616, -7, 167, 71, 4, -11, 522, 7377, -444, 626, 777, 11111, -72727]
        const test2 = ["Jack", "Alice", "Bob", "Jill", "Sue", "Fred", "Lue", "Zippy", "Scooter", "Lucy"]
        MergeSort.sort(test1)
        MergeSort.sort(test2)
        console.log(test1)
        console.log(test2)
    }
}

MergeSort.main()

File: Algorithms/Sorting/MergeSort/MergeSort.c
#include <stdio.h>

/* Prototypes */
void merge(int *arr, int *aux, int lo, int mid, int hi);
void sort(int *arr, int N);
void sort_helper(int *arr, int *aux, int lo, int hi);
void show_items(int *arr, int N);

int main(void)
{
    int test[18] = {6, 1, 7, 3, 4, 11, -66, 141, 664, -26, 1616, 2, 88, 1, 755, -2, -66, -2777};
    int N = 18;
    sort(test, N);
    show_items(test, N);
}

void sort(int *arr, int N)
{
    int aux[N];
    sort_helper(arr, aux, 0, N - 1);
}

void sort_helper(int *arr, int *aux, int lo, int hi)
{
    if (hi <= lo) { return; }
    int mid = lo + (hi - lo) / 2;
    sort_helper(arr, aux, lo, mid);
    sort_helper(arr, aux, mid + 1, hi);
    merge(arr, aux, lo, mid, hi);
}

void merge(int *arr, int *aux, int lo, int mid, int hi)
{
    // Copy arr into aux array
    for (int n = lo; n <= hi; n++) { aux[n] = arr[n]; }
    // Merge two halves
    int l = lo, r = mid + 1;
    for (int k = lo; k <= hi; k++)
    {
        if (l > mid) { arr[k] = aux[r++]; }
        else if (r > hi) { arr[k] = aux[l++]; }
        else if (aux[l] < aux[r]) { arr[k] = aux[l++]; }
        else { arr[k] = aux[r++]; }
    }
}

void show_items(int *arr, int N)
{
    printf("[");
    for (int i = 0; i < N; i++)
    {
        if (i == N - 1)
        {
            printf("%i]\n", arr[i]);
            break;
        }
        printf("%i, ", arr[i]);
    }
}
